src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php

```diff
@@ -176,43 +236,49 @@ public function indexAction(Request $request)
-        $form = $this->getForm($resource);
+        $this->isGrantedOr403($configuration, ResourceActions::CREATE);
 
-        if ($request->isMethod('POST') && $form->submit($request)->isValid()) {
-            $resource = $this->domainManager->create($form->getData());
+        $resource = $this->createNew($configuration);
+        $form = $this->createForm($configuration, $resource);
 
-            if ($this->config->isApiRequest()) {
-                if ($resource instanceof ResourceEvent) {
-                    throw new HttpException($resource->getErrorCode(), $resource->getMessage());
-                }
+        if ($request->isMethod('POST') && $form->submit($request)->isValid()) {
+            $this->eventDispatcher->dispatch(GenericResourceEvents::PRE_CREATE, new GenericResourceEvent($resource, $this->metadata, $configuration, ResourceActions::CREATE));
+            $event = $this->eventDispatcher->dispatchResourceEvent(ResourceEvents::PRE_CREATE, $resource);
 
-                return $this->handleView($this->view($resource, 201));
+            if ($event->isPropagationStopped() && !$configuration->isHtmlRequest()) {
+                throw new HttpException($event->getErrorCode(), $event->getMessage());
             }
 
-            if ($resource instanceof ResourceEvent) {
-                return $this->redirectHandler->redirectToIndex();
+            $this->manager->persist($resource);
+            $this->manager->flush();
+
+            $this->eventDispatcher->dispatch(GenericResourceEvents::POST_CREATE, new GenericResourceEvent($resource, $this->metadata, $configuration, ResourceActions::CREATE));
+            $this->eventDispatcher->dispatchResourceEvent(ResourceEvents::POST_CREATE, $resource);
+
+            if (!$configuration->isHtmlRequest()) {
+                return $this->viewHandler->handle(View::create($resource, 201));
             }
 
-            return $this->redirectHandler->redirectTo($resource);
+            return $this->redirectHandler->redirectToResource($configuration, $resource);
         }
 
-        if ($this->config->isApiRequest()) {
-            return $this->handleView($this->view($form, 400));
+        if (!$configuration->isHtmlRequest()) {
+            return $this->handleView($configuration, View::create($form, 400));
         }
 
-        $view = $this
-            ->view()
-            ->setTemplate($this->config->getTemplate('create.html'))
+        $view = View::create()
+            ->setTemplate($configuration->getTemplate(ResourceActions::CREATE.'.html'))
             ->setData(array(
-                $this->config->getResourceName() => $resource,
-                'form'                           => $form->createView(),
+                'metadata'                         => $this->metadata,
+                'resource'                         => $resource,
+                $this->metadata->getResourceName() => $resource,
+                'form'                             => $form->createView()
```

Comment:
Oh, I did not know about the ``setTemplateData`` method, that is awesome. Thanks!

Regarding your second comment - I wish there was a simple and clean way to do this. I will try to figure something out, but this is definitely for a separate controller.